astropy/modeling/core.py

```diff
@@ -3303,9 +3304,9 @@ def __getitem__(self, index):
                     if leftind == start and rightind == stop:
                         return node
                 raise IndexError("No appropriate subtree matches slice")
-        if isinstance(index, type(0)):
+        if isinstance(index, numbers.Integral):
```

Comment:
`numbers.Integral` will also allow boolean indexes though this should not be a stopper for the PR.